torch/nn/parallel/distributed.py

```diff
@@ -691,6 +728,52 @@ def __init__(
 
         self._setup_in_backward_optimizers()
 
+    def _register_delay_all_reduce_hook(
+        self,
+        bucket_cap_mb,
+        process_group,
+        param_to_hook_all_reduce,
+        device_ids,
+    ):
+        # 1. Create gradient buffer
+        device = torch.device("cpu") if device_ids is None else device_ids[0]
+        self._grad_buffer = torch.zeros(
+            sum([p.numel() for p in self._delay_all_reduce_params]),
+            device=device,
+        )
+
+        # 2. Broadcast the parameters
+        detached_params = [p.detach() for p in self._delay_all_reduce_params]
+        dist._broadcast_coalesced(process_group, detached_params, bucket_cap_mb, 0)
+
+        # 3. Hook all reduce to the specified parameter
+        world_size = dist.get_world_size(process_group)
+
+        def _delayed_all_reduce(grad):
+            self._grad_buffer.div_(world_size)
+            _ = dist.all_reduce(self._grad_buffer, group=process_group, async_op=True)
+            return grad
+
+        param_to_hook_all_reduce.register_hook(_delayed_all_reduce)
+
+        # 4. Build tensor views for gradients
+        offset = 0
+        for param in self._delay_all_reduce_params:
+            grad_view = self._grad_buffer[offset : (offset + param.numel())].view(
+                param.shape
+            )
+            self._grad_views.append(grad_view)
+            offset = offset + param.numel()
+
+        # 5. Check whether the all reduce of all params requiring grad is delayed.
+        for module_name, module in self.module.named_modules():
+            for param_name, param in module.named_parameters(recurse=False):
+                if param.requires_grad:
+                    full_name = f"{module_name}.{param_name}"
+                    if full_name not in self.parameters_to_ignore:
+                        return
```

Comment:
seems that it is not checking anything? it just simply return?

Reply:
If we have param that is not specified in parameters_to_ignore, we should not set _use_vanilla_module to True.